Add configurable logging to data directory

torkelrogstad opened the pull request:
Closes #588 

In this PR we add the ability for the node, chain
and wallet projects (+ the server) to log to the users
data directory instead of whatever directory the
binaries was launched from. This is inherently a bit
more complicated than our previous setup, because
we need to read the user directory before we can create
loggers. As a result of this, some files/methods were
moved around, so the relevant app config could be
found in scope.

## Discussion item: 
Right now we log in a similar way to how we already do it, by passing in the name of the file (instead of the class, but leads to almost identical behavior), and then constructing a logger with that name. This is why `sourcecode` (from Li Haoyi) is added as a dependency. Should we instead move to a category based system, where we log with a predefined set of categories? This is how both Bitcoin Core and Lnd handles it. Categories could be things such as chainsync, P2P messages, transaction creation, address derivation, peer handling, etc. IMO this is a better solution, primarily because when we have a predefined number of loggers its easy to expose options in `reference.conf` for fine-tuning behavior of each logger. I.e. the user could set P2P messages to be logged at `DEBUG`, but turn off chain sync logs. 

TODO: 
- [x] Make console/file logging something the user can turn on/off
- [x] Make specific loggers tunable wrt. levels
- [x] Add doc on logging for dev purposes

UPDATE: 

Added several logging categories, as well as options for tuning logging levels and file and console logs on/off.

torkelrogstad committed:
Add logging to data directory

In this commit we add the ability for the node, chain
and wallet projects (+ the server) to log to the users
data directory instead of whatever directory the
binaries was launched from. This is inherently a bit
more complicated than our previous setup, because
we need to read the user directory before we can create
loggers. As a result of this, some files/methods were
moved around, so the relevant app config could be
found in scope.

We also  introduce several logging categories that can be
tuned individually through user configuration. These logggers
are exposed both as traits that give a field `logger`, or
as methods that return the required logger.

nkohen commented on node/src/main/scala/org/bitcoins/node/peer/ConnectedPeerManager.scala:
```diff
@@ -1,10 +1,12 @@
 /** Manages all peers we have active connections with */
-class ConnectedPeerManager extends BitcoinSLogger {
+class ConnectedPeerManager extends P2PLogger {
+  implicit val config: ChainAppConfig = ???
```
ping

torkelrogstad replied:
This file isn't actually used anywhere

Christewart commented on chain/src/main/scala/org/bitcoins/chain/blockchain/sync/ChainSync.scala:
```diff
@@ -5,11 +5,12 @@ import org.bitcoins.chain.blockchain.ChainHandler
+import org.bitcoins.chain.config.ChainAppConfig
 
-trait ChainSync extends BitcoinSLogger {
+trait ChainSync {
```
`extends ChainVerificationLogger`?

Christewart commented on chain/src/main/scala/org/bitcoins/chain/blockchain/Blockchain.scala:
```diff
@@ -62,7 +61,9 @@ object Blockchain extends BitcoinSLogger {
-      implicit ec: ExecutionContext): Future[BlockchainUpdate] = {
+      implicit ec: ExecutionContext,
+      conf: ChainAppConfig): Future[BlockchainUpdate] = {
+    val logger = AppLoggers.getVerificationLogger
```
so why don't we extend the logger?

Christewart commented on db-commons/src/main/resources/reference.conf:
```diff
@@ -2,6 +2,36 @@ bitcoin-s {
+
+        # You can also tune specific module loggers.
+        # They each take the same levels as above.
+        # The available loggers are: 
```
can you give an example? For instance if i wanted to tune p2p would i do 

```
logging.p2p = INFO
```
?

Christewart commented on db-commons/src/main/scala/org/bitcoins/db/AppLoggers.scala:
```diff
@@ -0,0 +1,214 @@
+      case other              => sys.error(s"Expected LoggerContext, got: $other")
+    }
+
+    // following three lines prevents Logback from reading XML conf files
```
So does this mean our dev flow has to go through these loggers to? Or is this only applicable when the app is running?

torkelrogstad replied:
Both tests and running the app will log to the same place. By default that's both the console and the log file, so this doesn't really change development flow

Christewart commented on db-commons/src/main/resources/reference.conf:
```diff
@@ -2,6 +2,36 @@ bitcoin-s {
     datadir = ${HOME}/.bitcoin-s
     network = regtest # regtest, testnet3, mainnet
 
+    logging {
```
is there any way we can log everything to one file if we want to?

torkelrogstad replied:
WYM? Right now all loggers log to the same file

Christewart replied:
oh -- i thought they were being logged to different files in this PR. In the future they will be right? In that case, can we somehow having a setting to dump everything into one file?

torkelrogstad replied:
It's trivial to add new log files for specific loggers, as well as get one logger to output to several files if we want that in the future.

torkelrogstad committed:
Add datadir configuration to AppConfig

In this commit we add support for AppConfig to pick up
the data directory configuration file. We also add
a section to the contributing guide on how to tune
logging levels.

Christewart commented:
This fails to compile on 2.11

https://travis-ci.org/bitcoin-s/bitcoin-s/jobs/565913446#L621

Christewart commented:
Tests do not pass for me locally either 
```
[info] ChainAppConfigTest:
[info] - must be overridable *** FAILED ***
[info]   RegTest did not equal TestNet3 (ChainAppConfigTest.scala:21)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:530)
[info]   at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:529)
[info]   at org.scalatest.FlatSpec.newAssertionFailedException(FlatSpec.scala:1685)
[info]   at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:503)
[info]   at org.bitcoins.chain.ChainAppConfigTest.$anonfun$new$1(ChainAppConfigTest.scala:21)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
```

torkelrogstad commented:
The error you're having is not reproducing locally for me, nor on CI. Are you still seeing them, @Christewart ?

torkelrogstad committed:
Pass data directories explicitly for configuration

torkelrogstad commented:
I pushed a new commit that changes our app configs to explictly take in the data directory as a parameter, instead of magically resolving it. This also closes #588
